### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by Daniil-Pankieiev**

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,3 +1,19 @@
...
+    def wrapper(*args) -> Any:
+        key = args
+        if key in cached_results:
+            print("Getting from cache")
+            return cached_results[key]
+        else:
+            result = func(*args)
+            cached_results[key] = result
+            print("Calculating new result")
+            return result
```
Comment (LiudmylaKulinchenko) [inline app/main.py]: You can use a single return statement out of the if/else statement instead of two inside it.

File: app/main.py [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
-    pass
+from typing import Callable
+from typing import Any
+
+
+def cache(func: Callable) -> Callable:
+    cached_results = {}
+
+    def wrapper(*args) -> Any:
+        key = args
```
Comment (LiudmylaKulinchenko) [inline app/main.py]: Not really needed to create new variable here
Also, you can name the parameters in a function definition as you like, but it's a common convention to use `*args`. So, `args` will fit good here 👍

**Review by LiudmylaKulinchenko — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 1 above)
Comment (Daniil-Pankieiev) [reply app/main.py]: Fixed. Thanks it really looks better

**Comment by D1st3f** [comment]
All look like good!

**Review by D1st3f — APPROVED** [review]

File: app/main.py [hunk 3]
```diff
@@ -1,3 +1,17 @@
...
+    cached_results = {}
+
+    def wrapper(*args) -> Any:
+        if args in cached_results:
+            print("Getting from cache")
+            return cached_results[args]
+        result = func(*args)
+        cached_results[args] = result
+        print("Calculating new result")
+        return result
```
Comment (LiudmylaKulinchenko) [inline app/main.py]: Try rewrite with one return

**Review by LiudmylaKulinchenko — CHANGES_REQUESTED** [review]

**Review by wQuelS — CHANGES_REQUESTED** [review]
The solution looks quite good, but can you implement it without using the additional intermediate variable `result` ?

Also, fix conflicts that are shown on screenshot
image

**Review by wQuelS — APPROVED** [review]
gj

**Pull request closed without merging**